ipalib/x509.py

```diff
@@ -101,47 +104,353 @@ def strip_header(pem):
+        # but we need an iterable
+        extensions = self.__get_pyasn1_field('extensions') or []
+        OID_SAN = univ.ObjectIdentifier('2.5.29.17')
+        gns = []
+        for ext in extensions:
+            if ext['extnID'] == OID_SAN:
+                der = decoder.decode(
+                    ext['extnValue'], asn1Spec=univ.OctetString())[0]
+                gns = decoder.decode(der, asn1Spec=rfc2459.SubjectAltName())[0]
+                break
+        return gns
+
+    @property
+    def san_a_label_dns_names(self):
+        gns = self.__pyasn1_get_san_general_names()
+        result = []
+
+        for gn in gns:
+            if gn.getName() == 'dNSName':
+                result.append(unicode(gn.getComponent()))
+
+        return result
+
+    def match_hostname(self, hostname):
+        match_cert = {}
+
+        match_cert['subject'] = match_subject = []
+        for rdn in self._cert.subject.rdns:
+            match_rdn = []
+            for ava in rdn:
+                if ava.oid == crypto_x509.oid.NameOID.COMMON_NAME:
+                    match_rdn.append(('commonName', ava.value))
+            match_subject.append(match_rdn)
+
+        values = self.san_a_label_dns_names
+        if values:
+            match_cert['subjectAltName'] = match_san = []
+            for value in values:
+                match_san.append(('DNS', value))
+
+        ssl.match_hostname(match_cert, DNSName(hostname).ToASCII())
+
+
+def load_pem_x509_certificate(data):
+    """
+    Load an X.509 certificate in PEM format.
 
-    :param datatype: PEM for base64-encoded data (with or without header),
-                     or DER
-    :return: a python-cryptography ``CertificateSigningRequest`` object.
+    :returns: a ``_IPACertificate`` object.
     :raises: ``ValueError`` if unable to load the certificate.
+    """
+    return _IPACertificate(
+        crypto_x509.load_pem_x509_certificate(data, backend=default_backend())
+    )
 
+
+def load_der_x509_certificate(data):
     """
-    if type(data) in (tuple, list):
-        data = data[0]
+    Load an X.509 certificate in DER format.
 
-    if (datatype == PEM):
-        data = strip_header(data)
-        data = base64.b64decode(data)
+    :returns: a ``_IPACertificate`` object.
+    :raises: ``ValueError`` if unable to load the certificate.
+    """
+    return _IPACertificate(
+        crypto_x509.load_der_x509_certificate(data, backend=default_backend())
+    )
 
-    return cryptography.x509.load_der_x509_certificate(data, default_backend())
+
+def load_unknown_x509_certificate(data):
+    """
+    Only use this function when you can't be sure what kind of format does
+    your certificate have, e.g. input certificate files in installers
+
+    :returns: a ``_IPACertificate`` object.
+    :raises: ``ValueError`` if unable to load the certificate.
+    """
+    try:
+        return load_pem_x509_certificate(data)
+    except ValueError:
+        return load_der_x509_certificate(data)
```

Comment:
The old `x509.load_certificate(..., pem=True)` function does handle plain base64,
via `strip_headers()` which returns data unchanged if there are no headers.  The
stripped data then gets base64-decoded and parsed as DER.

I don't mind if we no longer support this, but neither will I be surprised when someone's
tests / deployment scripts / whatever breaks :)